### PR TITLE
feat(lep): customize maximum recurring job retain

### PR DESCRIPTION
**Pull request by mantissahz**
Ref: longhorn/longhorn#5713

### CODE REVIEW

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+## Proposal
+
+### User Stories
+
+Users can customize the maximum retain number when creating a recurring job to keep more snapshots or backups or restrict resources consumption especially disk storage.
+
+### User Experience In Detail
+
+Users can customize maximum retain number of the recurring job and maximum retain number cannot be exceeded when creating a recurring job.
```
Comment (innobead) [inline enhancements/20230821-customize-maximum-recurring-job-retain-number.md:31]: For snapshot retention, even though users can set a customized max retention number, it still needs to respect the factors which would impact snapshot creation, like the available space on disk, a new feature snapshot quota setting in 1.6, etc.

Please think about this.

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# Customize Maximum Recurring Job Retain Number
+
+## Summary
+
+Hardcoding the `MaxRecurringJobRetain` to be 100 may be inadequate in some cases such as that users try to keep 125 backups onn the remote backup target.
+This feature will make the `MaxRecurringJobRetain` to be customizable by users.
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md:6]: Added a NOTE section. Thanks

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md (hunk 1 above)
Comment (mantissahz) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md]: Added more description about potential impacts that users or system needs to respect when increasing the retain number. Thanks

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+Users can customize the maximum retain number when creating a recurring job to keep more snapshots or backups or restrict resources consumption especially disk storage.
+
+### User Experience In Detail
+
+Users can customize maximum retain number of the recurring job and maximum retain number cannot be exceeded when creating a recurring job.
+
+#### Kubectl
+
+User can set the global setting `recurring-job-max-retain` value to be a customized number under 250.
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md]: Modified. Thanks

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  name: recurring-job-max-retain
+  namespace: longhorn-system
+  resourceVersion: "87651"
+  uid: ab1f01cf-xxxx-45ab-9999-8ca1ebf61xxx
+value: "200"
+```
+
+#### Helm
+
+User can set `recurringJobMaxRetain` value to be a customized number under 250 such as:
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md:64]: Yes, 250 is our hard code upper bound for `maximumChainLength` in `longhorn-engine` and in [Snapshot space management LEP](https://github.com/longhorn/longhorn/pull/6634/files), snapshot max count is 250 still.

[the remark this answers is not in the document]
Comment (innobead) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md:64]: cc @FrankYang0529 @ChanYiLin

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+### API changes
+
+`None`
+
+## Design
+
+### Implementation Overview
+
+1. Add the global setting `recurring-job-max-retention` and default value is `100`.
+2. The number of global setting `recurring-job-max-retention` cannot be exceeded `250` because of the maximum number of snapshots for a volume.
```
Comment (innobead) [inline enhancements/20230821-customize-maximum-recurring-job-retain-number.md:84]: This setting will respect https://github.com/longhorn/longhorn/issues/6563 as well. If the limitation of snapshot is set but the retention exceeds the limitation, the maximum retention will be the snapshot limitation. 

This should be noted in the document.

**Review by innobead — APPROVED** [review]
In general, LGTM.

File: enhancements/20230821-customize-maximum-recurring-job-retain-number.md (hunk 5 above)
Comment (innobead) [reply enhancements/20230821-customize-maximum-recurring-job-retain-number.md:84]: cc @FrankYang0529 , is working on #6563 . Let's make sure the doc will be updated accordingly.

**Pull request merged**